### PR TITLE
Enabled MiniGrid gyms.

### PR DESCRIPTION
**Pull request by tatsubori**
Congratulations of this very very handy work making everyone happy.  :->

Just want to start discussing this quick hack enabling to use MiniGrid gyms.
https://github.com/maximecb/gym-minigrid
but not limited to this.

Instead of hard-coding imports, I have provided a command-line option to dynamically import specified gym packages for train.py and enjoy.py.

Adaptation of observation space (Dict) is hard-coded.

I am not sure if this breaks others, while I tried to be careful not to.

Cheers,
Mich

### CODE REVIEW

**Comment by araffin** [comment]
Hello,

Thank you for your interest.

I like the idea. Please don't forget to update the readme with the new instructions ;)

File: enjoy.py [hunk 1]
```diff
@@ -43,9 +43,13 @@ def main():
                         help='Normalize reward if applicable (trained with VecNormalize)')
     parser.add_argument('--seed', help='Random generator seed', type=int, default=0)
     parser.add_argument('--reward-log', help='Where to log reward', default='', type=str)
-
+    parser.add_argument('--gym-packages', type=str, nargs='+', help='Additional external Gym environemnt package modules to import (e.g. gym_minigrid)')
     args = parser.parse_args()
-
+    
+    for env_module in args.gym_packages:
+        import importlib
```
Comment (araffin) [inline enjoy.py]: Is there a reason why this import is not as the top ?

File: enjoy.py [hunk 2]
```diff
@@ -43,9 +43,13 @@ def main():
                         help='Normalize reward if applicable (trained with VecNormalize)')
     parser.add_argument('--seed', help='Random generator seed', type=int, default=0)
     parser.add_argument('--reward-log', help='Where to log reward', default='', type=str)
-
+    parser.add_argument('--gym-packages', type=str, nargs='+', help='Additional external Gym environemnt package modules to import (e.g. gym_minigrid)')
```
Comment (araffin) [inline enjoy.py]: this should add a default value, otherwise CI fails

File: enjoy.py [hunk 3]
```diff
@@ -43,9 +43,13 @@ def main():
...
     args = parser.parse_args()
-
+    
+    for env_module in args.gym_packages:
```
Comment (araffin) [inline enjoy.py]: Please add a comment

File: utils/utils.py [hunk 4]
```diff
@@ -78,6 +78,15 @@ def make_env(env_id, rank=0, seed=0, log_dir=None):
...
         set_global_seeds(seed + rank)
         env = gym.make(env_id)
+
+        # from baselines
+        #flatten_dict_observations = alg not in {'her'}
+        flatten_dict_observations = True
+        # check if it is a MiniGrid observation space
+        if flatten_dict_observations and isinstance(env.observation_space, gym.spaces.Dict):
+            keys = env.observation_space.spaces.keys()
+            env = gym.wrappers.FlattenDictWrapper(env, dict_keys=list(keys))
```
Comment (araffin) [inline utils/utils.py]: I did not know about that wrapper..., so from what I understand, it transform everything to a 1D vector?
In that case, the image is not normalized anymore...

Another solution, the goal env wrapper that will released where HER is ready: https://github.com/hill-a/stable-baselines/pull/273

File: hyperparams/ppo2.yml [hunk 5]
```diff
@@ -263,3 +263,33 @@ InvertedPendulumSwingupBulletEnv-v0:
...
   learning_rate: 2.5e-4
   cliprange: 0.2
+
+# following https://github.com/lcswillems/rl-starter-files
+# https://stable-baselines.readthedocs.io/en/master/modules/ppo2.html
+MiniGrid-DoorKey-5x5-v0:
+  normalize: true
+  n_envs: 8 # number of environment copies running in parallel
+  n_timesteps: !!float 1e5 # 80000 frames ?
+  policy: MlpPolicy # 'ActorCliticPolicy' or Mlp/Cnn/CnnLstmPolicy otherwise
```
Comment (araffin) [inline hyperparams/ppo2.yml]: I think you need to add a custom policy to handle the observation properly
See the architecture in: https://github.com/lcswillems/rl-starter-files

Also related: https://github.com/hill-a/stable-baselines/issues/321

File: hyperparams/ppo2.yml [hunk 6]
```diff
@@ -263,3 +263,33 @@ InvertedPendulumSwingupBulletEnv-v0:
   ent_coef: 0.0
   learning_rate: 2.5e-4
   cliprange: 0.2
+
+# following https://github.com/lcswillems/rl-starter-files
+# https://stable-baselines.readthedocs.io/en/master/modules/ppo2.html
+MiniGrid-DoorKey-5x5-v0:
```
Comment (araffin) [inline hyperparams/ppo2.yml]: btw, are those parameters working? If you could you add a trained agent to the repo?

File: enjoy.py (hunk 2 above)
Comment (tatsubori) [reply enjoy.py]: Thanks for your response and sorry for my dirty hacks.  ;->
Let me provide a smaller PR on this without a wrapper fallback for dict obs first, if complicated.

File: utils/utils.py (hunk 4 above)
Comment (tatsubori) [reply utils/utils.py]: Right.  I intended to provide a worst case fallback.  Will check on the other one.

File: hyperparams/ppo2.yml (hunk 5 above)
Comment (tatsubori) [reply hyperparams/ppo2.yml]: Encoding 2d observation image (even small one) + optional mission text into a vector might be too rough.  Would like to do it sooner or later.

File: hyperparams/ppo2.yml (hunk 6 above)
Comment (tatsubori) [reply hyperparams/ppo2.yml]: It is working as embedding everything into a vector for MLP.  Not super good compared to the one you mentioned, though.  CNN for partial map, Embedding/GRU for mission, MLPs for actor/critic should be the way to go...

Comment (tatsubori) [reply hyperparams/ppo2.yml]: If it is ok with naive models, they are ready.   :->

File: enjoy.py (hunk 1 above)
Comment (tatsubori) [reply enjoy.py]: Made the modification as much as local.   Moved to the top now.

**Comment by araffin** [comment]
The CI is failing because the docker image does not have minigrid installed... please add an exception in the test script so your trained models are not tested.

File: utils/utils.py [hunk 7]
```diff
@@ -78,6 +78,15 @@ def make_env(env_id, rank=0, seed=0, log_dir=None):
     def _init():
         set_global_seeds(seed + rank)
         env = gym.make(env_id)
+
+        # from baselines
+        #flatten_dict_observations = alg not in {'her'}
+        flatten_dict_observations = True
+        # check if it is a MiniGrid observation space
+        if flatten_dict_observations and isinstance(env.observation_space, gym.spaces.Dict):
```
Comment (araffin) [inline utils/utils.py]: I think it would be better to add an helper method for flattening dict obs, as it will be re-used for creating a test env.

File: utils/utils.py (hunk 4 above)
Comment (tatsubori) [reply utils/utils.py]: I have just read through the thread.  Right now it requires "all keys to have the same type", which is not the case, as it is unfortunately.  Need to look into it further to understand its ideas.

File: hyperparams/ppo2.yml (hunk 5 above)
Comment (tatsubori) [reply hyperparams/ppo2.yml]: I have just read through the Stable Baselines conversation.
So, we should take the vectorize-it-and-reshape-it approach.  Got it.

File: utils/utils.py (hunk 7 above)
Comment (tatsubori) [reply utils/utils.py]: Done.  In the same file for now, as flatten_dict_observations(env).
Feel free to tell your favor in code styles if any BTW.

File: hyperparams/ppo2.yml (hunk 5 above)
Comment (tatsubori) [reply hyperparams/ppo2.yml]: Though custom policies are of my interest (to have the similar architecture to rl-starter-files), it is not quick to complete today.  Let me do this next week, hopefully as another PR.  :->
Your insight has been really helpful.  Thanks!

**Comment by tatsubori** [comment]
> The CI is failing because the docker image does not have minigrid installed... please add an exception in the test script so your trained models are not tested.

Added --gym-packages gym_minigrid if "-MiniGrid-" in trained_model.
pip install gym-minigrid for docker scripts (cpu & gpu) as well.

File: hyperparams/ppo2.yml (hunk 5 above)
Comment (araffin) [reply hyperparams/ppo2.yml]: your welcome ;)

**Comment by araffin** [comment]
Good, I'll test that on my local machine before pushing a new docker image (so the CI can pass)

**Comment by araffin** [comment]
I got qt segmentation fault on my machine trying to render the scene :/
Anyway, it seems ok otherwise.

I would also use the flat obs wrapper from minigrid: https://github.com/maximecb/gym-minigrid/blob/master/gym_minigrid/wrappers.py#L167

**Comment by tatsubori** [comment]
Better to exclude the cases from the test case for now?

**Comment by araffin** [comment]
>Better to exclude the cases from the test case for now?

yep! (that was my first remark ;))

**Comment by araffin** [comment]
@tatsubori is that ok if I merge now and in another PR, you'll include a better support for minigrid? (using FlatObsWrapper for instance)

**Comment by tatsubori** [comment]
Yes, please proceed to merge.  I have CNN impl with it at hand but it won’t converge.  I am trying CNN+Lstm but it may take time.

Will continue enhancing it.

StableBaselines to support (not to fail) Dict is better to handle mission text, I believe, though.

**Comment by araffin** [comment]
>StableBaselines to support (not to fail)

I agree, but the support is far from trivial ;) (however we would appreciate PR to solve that issue)
